File: computer_science/data_structures/stack/stack.py
# Stack Big O complexity

# Push: O(1) - Constant Time
# Pop (remove): O(1) - Constant Time
# top (top): O(1) - Constant Time
# Is Empty: O(1) - Constant Time
# Size: O(1) - Constant Time


class Emptiness(Exception):
    pass


class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if self.is_empty():
            raise Emptiness('The Stack is empty')

        return self.items.pop()

    def is_empty(self):
        return self.size() == 0

    def top(self):
        if self.is_empty():
            raise Emptiness('The Stack is empty')

        return self.items[-1]

    def size(self):
        return len(self.items)


stack = Stack()

print(stack.is_empty())

stack.push(1)
print(stack.items)
stack.push(2)
print(stack.items)
stack.push(3)
print(stack.items)
stack.push(4)
print(stack.items)
stack.push(5)
print(stack.items)

print(stack.is_empty())
print(stack.top())

print(stack.pop())
print(stack.pop())
print(stack.pop())
print(stack.pop())

print(stack.is_empty())

print(stack.pop())

print(stack.is_empty())

# reversing a list


def reverse(bookshelf):
    stack = Stack()

    for book in bookshelf:
        stack.push(book)

    reversed_bookshelf = []

    while not stack.is_empty():
        reversed_bookshelf.append(stack.pop())

    return reversed_bookshelf


bookshelf = [
    'Harry Potter',
    'Atomic Habits',
    'Leonardo da Vinci',
    'Sapiens',
    'Peak'
]

reversed_bookshelf = reverse(bookshelf)

print(reversed_bookshelf)

File: computer_science/data_structures/stack/stack.js
const push = (stack, item) => [item, ...stack];

const pop = (stack) => stack.slice(1);

const top = (stack) => stack[0];

const isEmpty = (stack) => size(stack) === 0;

const size = (stack) => stack.length;

// testing
let stack = [];

console.log(isEmpty(stack));

stack = push(stack, 1);
console.log(stack);
stack = push(stack, 2);
console.log(stack);
stack = push(stack, 3);
console.log(stack);
stack = push(stack, 4);
console.log(stack);
stack = push(stack, 5);
console.log(stack);

console.log(top(stack));
console.log(size(stack));
console.log(isEmpty(stack));

stack = pop(stack);
stack = pop(stack);
stack = pop(stack);
stack = pop(stack);

console.log(isEmpty(stack));
console.log(top(stack));
stack = pop(stack);
console.log(isEmpty(stack));
console.log(size(stack));
